File: bin/json_to_parquet.py
#!/usr/bin/env python3
"""Convert the JSON dataset into the Parquet format used by the app."""

from __future__ import annotations

import argparse
import json
import uuid
from collections.abc import Iterable
from pathlib import Path
from typing import Any, TypedDict, cast

import pyarrow as pa
import pyarrow.parquet as pq

DEFAULT_JSON_PATH = Path("static/data.json")
DEFAULT_PARQUET_PATH = Path("static/data.parquet")


class CreatedBy(TypedDict, total=False):
    name: str | None
    website: str | None


class NormalisedEntry(TypedDict):
    id: str
    word: str
    definition: str
    example: str
    createdBy: CreatedBy
    createdAt: str


def load_json(json_path: Path) -> list[dict[str, Any]]:
    with json_path.open("r", encoding="utf-8") as handle:
        data: Any = json.load(handle)
    if not isinstance(data, list):
        raise RuntimeError("Dataset JSON must be a list of entries")

    entries: list[dict[str, Any]] = []
    for item in data:
        if not isinstance(item, dict):
            raise RuntimeError("Dataset JSON entries must be objects")
        entries.append(cast(dict[str, Any], item))
    return entries


def _coerce_entry(raw: dict[str, Any]) -> NormalisedEntry:
    def _to_str(value: Any) -> str:
        if isinstance(value, str):
            return value
        return str(value) if value is not None else ""

    def _to_optional_str(value: Any) -> str | None:
        text = _to_str(value)
        return text if text else None

    raw_id = raw.get("id")
    if isinstance(raw_id, str) and raw_id.strip():
        entry_id = raw_id.strip()
    else:
        entry_id = str(uuid.uuid4())

    if "word" in raw:
        created_by = raw.get("createdBy") or {}
        return {
            "id": entry_id,
            "word": _to_str(raw.get("word")),
            "definition": _to_str(raw.get("definition", "")),
            "example": _to_str(raw.get("example", "")),
            "createdBy": {
                "name": _to_optional_str(created_by.get("name")),
                "website": _to_optional_str(created_by.get("website")),
            },
            "createdAt": _to_str(raw.get("createdAt", "")),
        }

    synonyms = raw.get("synonyms") or []
    cleaned_synonyms = ", ".join(
        _to_str(s).strip() for s in synonyms if _to_str(s).strip()
    )
    definition = _to_str(raw.get("meaning", "")).strip()
    if cleaned_synonyms:
        definition = f"{definition}\n\nSinónimos: {cleaned_synonyms}"

    authors = raw.get("authors") or []
    author_names = ", ".join(
        _to_str(a.get("name")).strip()
        for a in authors
        if _to_str(a.get("name")).strip()
    )
    author_websites = ", ".join(
        _to_str(a.get("link")).strip()
        for a in authors
        if _to_str(a.get("link")).strip()
    )

    return {
        "id": entry_id,
        "word": _to_str(raw.get("text", "")),
        "definition": definition,
        "example": "\n".join(
            _to_str(ex).strip("\n") for ex in (raw.get("examples") or [])
        ),
        "createdBy": {
            "name": author_names or None,
            "website": author_websites or None,
        },
        "createdAt": _to_str(raw.get("createdAt", "2021-08-31T00:00:00.000Z")),
    }


def transform_entries(data: Iterable[dict[str, Any]]) -> list[NormalisedEntry]:
    transformed: list[NormalisedEntry] = []
    for entry in data:
        transformed.append(_coerce_entry(entry))
    return transformed


def write_parquet(entries: list[NormalisedEntry], parquet_path: Path) -> None:
    created_by_struct = pa.struct(
        [
            pa.field("name", pa.string()),
            pa.field("website", pa.string()),
        ]
    )
    schema = pa.schema(
        [
            pa.field("id", pa.string()),
            pa.field("word", pa.string()),
            pa.field("definition", pa.string()),
            pa.field("example", pa.string()),
            pa.field("createdBy", created_by_struct),
            pa.field("createdAt", pa.string()),
        ]
    )

    table = pa.Table.from_pydict(
        {
            "id": [entry["id"] for entry in entries],
            "word": [entry["word"] for entry in entries],
            "definition": [entry["definition"] for entry in entries],
            "example": [entry["example"] for entry in entries],
            "createdBy": pa.array(
                [entry["createdBy"] for entry in entries],
                type=created_by_struct,
            ),
            "createdAt": [entry["createdAt"] for entry in entries],
        },
        schema=schema,
    )

    parquet_path.parent.mkdir(parents=True, exist_ok=True)
    pq.write_table(table, parquet_path, version="2.6", compression="snappy")


def save_json(entries: list[NormalisedEntry], json_path: Path) -> None:
    json_path.write_text(
        json.dumps(entries, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--json",
        type=Path,
        default=DEFAULT_JSON_PATH,
        help=f"Source JSON file (default: {DEFAULT_JSON_PATH})",
    )
    parser.add_argument(
        "--parquet",
        type=Path,
        default=DEFAULT_PARQUET_PATH,
        help=f"Destination Parquet file (default: {DEFAULT_PARQUET_PATH})",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    data = load_json(args.json)
    entries = transform_entries(data)
    save_json(entries, args.json)
    write_parquet(entries, args.parquet)
    print(f"✅ Migrated {len(entries)} entries from {args.json} -> {args.parquet}")


if __name__ == "__main__":
    main()

File: scripts/json_to_parquet.js
#!/usr/bin/env node
/**
 * Convert the JSON dataset into the Parquet format used by the app.
 */

import { readFileSync, writeFileSync, mkdirSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { randomUUID, createHash } from "crypto";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const DEFAULT_JSON_PATH = resolve(__dirname, "../data.json");
const DEFAULT_PARQUET_PATH = resolve(__dirname, "../static/data.parquet");
const DEFAULT_METADATA_PATH = resolve(__dirname, "../static/data.parquet.json");

/**
 * CreatedBy structure
 * @typedef {Object} CreatedBy
 * @property {string} name - Author name (required, non-nullable)
 * @property {string} website - Author website (required, non-nullable)
 */

/**
 * Entry structure
 * @typedef {Object} NormalizedEntry
 * @property {string} id - Unique identifier
 * @property {string} word - Word or expression
 * @property {string} definition - Definition
 * @property {string} example - Usage example
 * @property {CreatedBy} createdBy - Author information (single author)
 * @property {string} createdAt - ISO timestamp
 */

/**
 * Load JSON data
 * @param {string} jsonPath - Path to JSON file
 * @returns {any[]} Array of entries
 */
function loadJson(jsonPath) {
  const data = JSON.parse(readFileSync(jsonPath, "utf-8"));
  if (!Array.isArray(data)) {
    throw new Error("Dataset JSON must be a list of entries");
  }
  return data;
}

/**
 * Convert value to string
 * @param {any} value - Value to convert
 * @returns {string} String representation
 */
function toStr(value) {
  if (typeof value === "string") return value;
  return value != null ? String(value) : "";
}

/**
 * Coerce raw entry to normalized format
 * @param {any} raw - Raw entry data
 * @returns {NormalizedEntry} Normalized entry
 */
function coerceEntry(raw) {
  const rawId = raw.id;
  let entryId;
  if (typeof rawId === "string" && rawId.trim()) {
    entryId = rawId.trim();
  } else {
    entryId = randomUUID();
  }

  if ("word" in raw) {
    const createdBy = raw.createdBy || {};
    return {
      id: entryId,
      word: toStr(raw.word),
      definition: toStr(raw.definition || ""),
      example: toStr(raw.example || ""),
      createdBy: {
        name: toStr(createdBy.name || "Anónimo"),
        website: toStr(createdBy.website || ""),
      },
      createdAt: toStr(raw.createdAt || ""),
    };
  }

  // Handle old format with synonyms and single author
  const synonyms = raw.synonyms || [];
  const cleanedSynonyms = synonyms
    .map((s) => toStr(s).trim())
    .filter((s) => s)
    .join(", ");

  let definition = toStr(raw.meaning || "").trim();
  if (cleanedSynonyms) {
    definition = `${definition}\n\nSinónimos: ${cleanedSynonyms}`;
  }

  const authors = raw.authors || [];
  const firstAuthor = authors[0] || {};
  const authorName = toStr(firstAuthor.name || "Anónimo").trim();
  const authorWebsite = toStr(firstAuthor.link || "").trim();

  return {
    id: entryId,
    word: toStr(raw.text || ""),
    definition,
    example: (raw.examples || []).map((ex) => toStr(ex).trim()).join("\n"),
    createdBy: {
      name: authorName || "Anónimo",
      website: authorWebsite,
    },
    createdAt: toStr(raw.createdAt || "2021-08-31T00:00:00.000Z"),
  };
}

/**
 * Transform entries to normalized format
 * @param {any[]} data - Array of raw entries
 * @returns {NormalizedEntry[]} Array of normalized entries
 */
function transformEntries(data) {
  return data.map((entry) => coerceEntry(entry));
}

/**
 * Write parquet file
 * @param {NormalizedEntry[]} entries - Array of normalized entries
 * @param {string} parquetPath - Path to parquet file
 * @returns {Promise<void>}
 */
async function writeParquet(entries, parquetPath) {
  const { parquetWriteFile } = await import("hyparquet-writer");

  mkdirSync(dirname(parquetPath), { recursive: true });

  const columnData = [
    { name: "id", data: entries.map((e) => e.id), type: "STRING" },
    { name: "word", data: entries.map((e) => e.word), type: "STRING" },
    { name: "definition", data: entries.map((e) => e.definition), type: "STRING" },
    { name: "example", data: entries.map((e) => e.example), type: "STRING" },
    { name: "createdByName", data: entries.map((e) => e.createdBy.name), type: "STRING" },
    { name: "createdByWebsite", data: entries.map((e) => e.createdBy.website), type: "STRING" },
    { name: "createdAt", data: entries.map((e) => e.createdAt), type: "STRING" },
  ];

  await parquetWriteFile({
    filename: parquetPath,
    columnData,
  });
}

/**
 * Save normalized JSON
 * @param {NormalizedEntry[]} entries - Array of normalized entries
 * @param {string} jsonPath - Path to JSON file
 * @returns {void}
 */
function saveJson(entries, jsonPath) {
  writeFileSync(jsonPath, JSON.stringify(entries, null, 2), "utf-8");
}

/**
 * Generate metadata for parquet file
 * @param {string} parquetPath - Path to parquet file
 * @param {string} metadataPath - Path to metadata file
 * @returns {void}
 */
function generateMetadata(parquetPath, metadataPath) {
  const fileBuffer = readFileSync(parquetPath);
  const hash = createHash("sha256").update(fileBuffer).digest("hex");

  const metadata = {
    hash,
    version: "1.0",
    createdAt: new Date().toISOString(),
    fileSize: fileBuffer.length,
  };

  writeFileSync(metadataPath, JSON.stringify(metadata, null, 2), "utf-8");
  console.log(
    `✅ Generated metadata: hash=${hash.substring(0, 12)}..., size=${(fileBuffer.length / 1024 / 1024).toFixed(2)} MB`
  );
}

async function main() {
  const data = loadJson(DEFAULT_JSON_PATH);
  const entries = transformEntries(data);
  saveJson(entries, DEFAULT_JSON_PATH);
  await writeParquet(entries, DEFAULT_PARQUET_PATH);
  console.log(
    `✅ Migrated ${entries.length} entries from ${DEFAULT_JSON_PATH} -> ${DEFAULT_PARQUET_PATH}`
  );
  generateMetadata(DEFAULT_PARQUET_PATH, DEFAULT_METADATA_PATH);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});
